### DISCUSSION TITLE
OSDs not coming up. Controller not reconciling

### DISCUSSION THREAD

**Discussion by ilveroluca** [rook/rook#13285] [Q&A]
Hi,

I'm trying to resurrect a Rook-Ceph cluster. The cluster broke after an outage.

I was able to bring the mons back up and establish quorum by removing the existing mon-endpoints ConfigMap, which had stale data. The controller then restarted the mons, created a correct endpoints ConfigMap, and then was able to restart most of the dependent services (e.g., mgr, mds, etc.).

My problem is that the OSDs aren't coming back up. The OSD pods are in CrashLoopBackOff state; when trying to start they log the following:
```
debug 2023-11-24T16:00:42.088+0000 7ff5a214c700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2]
debug 2023-11-24T16:00:42.088+0000 7ff5a114a700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2]
debug 2023-11-24T16:00:42.088+0000 7ff5a194b700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2]
failed to fetch mon config (--no-mon-config to skip)
```
I tried to manually launch an OSD, but I wasn't able to do so.  I created a debug pod by patching one of the OSD pods to run sleep and opening a shell in it. Manually launching the OSD, using the same command specified in the deployment, naturally produces the same error.   I did notice that a simple `ceph status` in the same pod fails to contact the cluster unless I also provide a `--conf` as in:
```
ceph  --conf /var/lib/rook/rook-ceph/rook-ceph.config status
  cluster:
    id:     1bb8caa8-d8d6-42cf-957e-9611951303c0
    health: HEALTH_WARN
....
```
(note: `rook-ceph` is the name of the cluster). That leads me to suspect that maybe the operator isn't creating or mounting the required configuration?

Trying to manully provide the same config file to the ceph-osd command (with `--conf ...`) results in messages about ceph-osd not being to find the keyring file (or rather, it's looking for it under at `/var/lib/ceph/osd/rook-ceph-0/keyring`, which to me suggests that the rook-ceph.config file, while syntactically compatible, is not the right one for ceph-osd).

The operator itself is "stuck", as in it continuously spits out the message
```
│ 2023-11-24 17:04:47.167463 I | op-mon: parsing mon endpoints: d=10.233.11.39:6789,a=10.233.14.157:6789,e=10.233.38.74:6789
```
It now seems to have given up on its reconciliation tasks, given the messages like this:
```
 2023-11-24 16:58:11.578216 I | ceph-spec: ceph-object-controller: CephCluster "rook-ceph" found but skipping reconcile since ceph health is &{"HEALTH_ERR" map["error":{"Urgent" "failed to get status. . Error initializing cluster client: ObjectNotFound('RADOS object not found (error calling conf_read_file)',): exit status 1"}] "2023-11-24T16:57:11Z" "2023-11-24T16:57:11Z" "HEALTH_WARN"}
```

I'm stuck too now.  I tried deleting the OSD deployments, hoping that might trigger action from the controller, but it didn't.

Can anyone suggest how I may be able to get this going again?

**Comment by travisn** [comment]
By deleting the mon config, you may have reset the mons to start a new cluster. In that case the OSDs wouldn't be able to launch. If the same cluster fsid is still in use, then it sounds like the OSDs don't have the correct auth. You could attempt to update the `ceph osd auth` for each OSD to use the auth that the OSDs expect. But if the mons reset to use a different fsid then I'm not sure the OSDs will be able to connect again since it's like OSDs from another cluster trying to connect, which will of course be denied access.

**Reply by ilveroluca** [reply]
Thanks for your help thus far. I had saved a definition of an OSD deployment before deleting them.  I used it to recreate an OSD, applied to same "sleep" trick as before to open a shell in the OSD pod.  

Here's what I found:
```
$ kl -n rook-ceph exec -it deploy/rook-ceph-osd-0 -- bash
[root@rook-ceph-osd-0-6c656bfb89-jcx7q /]# cd /var/lib/ceph/osd/ceph-0/
[root@rook-ceph-osd-0-6c656bfb89-jcx7q ceph-0]# ls
block  ceph_fsid  fsid  keyring  ready  type  whoami
[root@rook-ceph-osd-0-6c656bfb89-jcx7q ceph-0]# cat ceph_fsid 
1bb8caa8-d8d6-42cf-957e-9611951303c0
```
The content of ceph_fsid found in the OSD-0 pod is the same as the fsid reported by `ceph status` executed in the rook-ceph-tools pod.  
```
$ kl -n rook-ceph exec -it deploy/rook-ceph-tools -- ceph status
  cluster:
    id:     1bb8caa8-d8d6-42cf-957e-9611951303c0
    health: HEALTH_WARN
...
```
Is that sufficient to claim that the fsid hasn't changed?

That would leave the OSD authentication problem.

From inside the OSD-0 pod, I was going to look into how to use `ceph osd auth`.  I ran `ceph osd auth --help` and found that the command times out:
```
[root@rook-ceph-osd-0-6c656bfb89-jcx7q ceph-0]# ceph osd auth --help

 General usage:
 ==============
usage: ceph [-h] [-c CEPHCONF]....
<snip>
Local commands:
 ===============

ping <mon.id>           Send simple presence/life test to a mon
                        <mon.id> may be 'mon.*' for all mons
daemon {type.id|path} <cmd>
                        Same as --admin-daemon, but auto-find admin socket
daemonperf {type.id | path} [stat-pats] [priority] [<interval>] [<count>]
daemonperf {type.id | path} list|ls [stat-pats] [priority]
                        Get selected perf stats from daemon/admin socket
                        Optional shell-glob comma-delim match string stat-pats
                        Optional selection priority (can abbreviate name):
                         critical, interesting, useful, noninteresting, debug
                        List shows a table of all available stats
                        Run <count> times (default forever),
                         once per <interval> seconds (default 1)

timed out
[root@rook-ceph-osd-0-6c656bfb89-jcx7q ceph-0]#
```
While I was at it, I tried executing some other read-only commands from the same pod. e.g. `ceph auth ls`, `ceph auth print-key`, `ceph fsid`, `ceph osd blocklist ls`. They all hang. Killing them I get the message `Cluster Connection aborted`.

Does this tell us anything useful? How do the ceph commands know where to find the cluster?  I see the `CEPH_ARGS` variable is set:
```
# echo $CEPH_ARGS
-m [v2:10.233.38.74:3300,v1:10.233.38.74:6789],[v2:10.233.11.39:3300,v1:10.233.11.39:6789],[v2:10.233.14.157:3300,v1:10.233.14.157:6789]
```
The values match the addresses of three of the mon ClusterIP services.

Anything else I can check?

**Reply by travisn** [reply]
Good to see that the fsid did not change. 

To check the OSD auth, try this [past comment on updating the OSD auth](https://github.com/rook/rook/issues/4238#issuecomment-664627282).

**Reply by ilveroluca** [reply]
Thanks for your help @travisn. I applied your suggestion. Incidentally, I realized that rook puts the keyrings in kubernetes secrets, so they can be retrieved from there rather than mucking with the pods.

Once I imported they keyrings, the OSDs managed to boot.  Something's still wrong though.

`ceph status` reports 5 osds, as expected, but 0 are up and so 100% of pgs are still unavailable:
```
  services:
    mon: 3 daemons, quorum a,d,e (age 105m)
    mgr: a(active, since 9m)
    mds: tdmfs:1 {0=tdmfs-a=up:creating} 1 up:standby-replay
    osd: 5 osds: 0 up, 0 in

  data:
    pools:   11 pools, 177 pgs
    objects: 0 objects, 0 B
    usage:   0 B used, 0 B / 0 B avail
    pgs:     100.000% pgs unknown
             177 unknown
```
Running `ceph osd tree`  shows all OSDs down.

The logs from the OSD prepare pods show that they found the LVM storage device:
```
2023-12-05 17:15:51.876236 I | cephosd: no new devices to configure. returning devices already configured with ceph-volume.
2023-12-05 17:15:51.876448 D | exec: Running command: stdbuf -oL ceph-volume --log-path /tmp/ceph-log lvm list  --format json
2023-12-05 17:15:52.328812 D | cephosd: {
    "0": [
        {
            "devices": [
                "/dev/vdb"
            ],
            "lv_name": "osd-data-83732cba-3d09-4133-bcd8-eade54fbe9a9",
            "lv_path": "/dev/ceph-bf932c12-e6fb-42b7-91cb-ade1ee6aabb4/osd-data-83732cba-3d09-4133-bcd8-eade54fbe9a9",
            "lv_size": "7.71t",
...
```
I see that the `ceph.cluster_fsid` tag value matches the fsid. 

The `activate` initContainer from the OSD pod reports no error and I see that it logs
```
--> ceph-volume lvm activate successful for osd ID: 0
```
As the OSDs initialize, they don't seem to report any errors. On the bright side, they report finding the PGS. There's this message about metadata for the storage device that may be relevant:
```
debug 2023-12-05T16:51:42.580+0000 7f6251eb7700  1 osd.0 568825 _collect_metadata vdb:  no unique device id for vdb: fallback method has no model nor serial'
```
Also, this warning seems to be negative:
```
debug 2023-12-05T16:51:23.068+0000 7f625e1b0f40  1 bluestore(/var/lib/ceph/osd/ceph-0) _upgrade_super done
debug 2023-12-05T16:51:23.412+0000 7f625e1b0f40  0 _get_class not permitted to load sdk
debug 2023-12-05T16:51:23.444+0000 7f625e1b0f40  0 _get_class not permitted to load queue
debug 2023-12-05T16:51:23.460+0000 7f625e1b0f40  0 <cls> /home/jenkins-build/build/workspace/ceph-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/15.2.4/rpm/el8/BUILD/ceph-15.2.4/src/cls/cephfs/cls_cephfs.cc:198: loading cephfs
debug 2023-12-05T16:51:23.464+0000 7f625e1b0f40  0 <cls> /home/jenkins-build/build/workspace/ceph-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/15.2.4/rpm/el8/BUILD/ceph-15.2.4/src/cls/hello/cls_hello.cc:312: loading cls_hello
debug 2023-12-05T16:51:23.464+0000 7f625e1b0f40  0 _get_class not permitted to load kvs
debug 2023-12-05T16:51:23.464+0000 7f625e1b0f40  0 _get_class not permitted to load lua
debug 2023-12-05T16:51:23.464+0000 7f625e1b0f40  1 osd.0 568825 warning: got an error loading one or more classes: (1) Operation not permitted
```
After they initialize, the OSDs log this frequently:
```
debug 2023-12-05T17:36:27.692+0000 7f6254f2d700  1 osd.0 568825 tick checking mon for new map
```
And then, every so often, there's a stats dump from rocks db.

The mons, on the other hand, are steadily logging these attempts to retrieve OSD metadata by the mgr:
```
audit 2023-12-06T08:08:34.219896+0000 mon.e (mon.2) 282767 : audit [DBG] from='mgr.1286783 10.233.124.174:0/1685082539' entity='mgr.a' cmd=[{"prefix": "osd metadata", "id": 3}]: dispatch
```
And the mgr reports messages like these (looks like one per PGS per OSD), which don't look ok:
```
debug 2023-12-05T17:36:22.144+0000 7f8e078ed700  0 log_channel(cluster) log [DBG] : pgmap v173: 177 pgs: 177 unknown; 0 B data, 0 B used, 0 B / 0 B avail
debug 2023-12-05T17:36:23.036+0000 7f8e23939700  1 mgr finish mon failed to return metadata for osd.4: (2) No such file or directory
debug 2023-12-05T17:36:23.036+0000 7f8e23939700  1 mgr finish mon failed to return metadata for osd.2: (2) No such file or directory
debug 2023-12-05T17:36:23.036+0000 7f8e23939700  1 mgr finish mon failed to return metadata for osd.3: (2) No such file or directory
...
```
I'm attaching part of the log files, in case someone can spot something more informative.

Any other suggestions to help me resurrect this cluster?

[mgr-a.log](https://github.com/rook/rook/files/13579326/mgr-a.log)
[mon-a_head.log](https://github.com/rook/rook/files/13579327/mon-a_head.log)
[osd-0.log](https://github.com/rook/rook/files/13579328/osd-0.log)

**Reply by JosephWalter10** [reply]
Can someone please help me fix the issue?

**Reply by sp98** [reply]
@JosephWalter10 This is discussion from 2023. What issues are you facing?

**Reply by JosephWalter10** [reply]
**Logs of Monitor Pod:**

debug 2026-02-17T09:47:10.136+0000 7fc7737d4700  0 log_channel(audit) log [DBG] : from='mgr.118161 10.233.64.172:0/1114054608' entity='mgr.a' cmd=[{"prefix": "osd metadata", "id": 0}]: dispatch
debug 2026-02-17T09:47:10.156+0000 7fc7737d4700  0 mon.c@0(leader) e22 handle_command mon_command({"prefix": "osd metadata", "id": 4} v 0) v1
debug 2026-02-17T09:47:10.156+0000 7fc7737d4700  0 log_channel(audit) log [DBG] : from='mgr.118161 10.233.64.172:0/1114054608' entity='mgr.a' cmd=[{"prefix": "osd metadata", "id": 4}]: dispatch
debug 2026-02-17T09:47:10.316+0000 7fc7737d4700  0 mon.c@0(leader) e22 handle_command mon_command({"prefix": "osd metadata", "id": 7} v 0) v1
debug 2026-02-17T09:47:10.316+0000 7fc7737d4700  0 log_channel(audit) log [DBG] : from='mgr.118161 10.233.64.172:0/1114054608' entity='mgr.a' cmd=[{"prefix": "osd metadata", "id": 7}]: dispatch
debug 2026-02-17T09:47:10.340+0000 7fc7737d4700  0 mon.c@0(leader) e22 handle_command mon_command({"prefix": "osd tree", "format": "json"} v 0) v1
debug 2026-02-17T09:47:10.340+0000 7fc7737d4700  0 log_channel(audit) log [DBG] : from='client.? 10.233.64.198:0/842599139' entity='client.admin' cmd=[{"prefix": "osd tree", "format": "json"}]: dispatch
debug 2026-02-17T09:47:10.356+0000 7fc7737d4700  0 mon.c@0(leader) e22 handle_command mon_command({"prefix": "osd metadata", "id": 3} v 0) v1

**Logs of Manager Pod:**

debug 2026-02-17T09:51:23.991+0000 7fc878417700  1 mgr finish mon failed to return metadata for osd.5: (2) No such file or drectory
debug 2026-02-17T09:51:24.179+0000 7fc6c3668700  2 mgr.server handle_open ignoring open from osd.0 v2:10.15.15.7:6800/361546812; not ready for session (expect reconnect)
debug 2026-02-17T09:51:24.179+0000 7fc878c18700  1 mgr finish mon failed to return metadata for osd.0: (2) No such file or drectory
debug 2026-02-17T09:51:24.195+0000 7fc6c3668700  2 mgr.server handle_open ignoring open from osd.4 v2:10.15.15.29:6800/361650697; not ready for session (expect reconnect)
debug 2026-02-17T09:51:24.199+0000 7fc878417700  1 mgr finish mon failed to return metadata for osd.4: (2) No such file or drectory

**Logs of OSD Pod:**

Cumulative compaction: 0.11 GB write, 0.00 MB/s write, 0.11 GB read, 0.00 MB/s read, 0.9 seconds
Interval compaction: 0.00 GB write, 0.00 MB/s write, 0.00 GB read, 0.00 MB/s read, 0.0 seconds
Stalls(count): 0 level0_slowdown, 0 level0_slowdown_with_compaction, 0 level0_numfiles, 0 level0_numfiles_with_compaction, 0 stop for pending_compaction_bytes, 0 slowdown for pending_compaction_bytes, 0 memtable_compaction, 0 memtable_slowdown, interval 0 total count

** File Read Latency Histogram By Level [P] **

debug 2026-02-17T09:47:30.402+0000 7f44b3f23700  1 osd.0 1216 tick checking mon for new map
debug 2026-02-17T09:48:00.575+0000 7f44b3f23700  1 osd.0 1216 tick checking mon for new map
debug 2026-02-17T09:48:30.699+0000 7f44b3f23700  1 osd.0 1216 tick checking mon for new map
debug 2026-02-17T09:49:01.679+0000 7f44b3f23700  1 osd.0 1216 tick checking mon for new map
debug 2026-02-17T09:49:32.543+0000 7f44b3f23700  1 osd.0 1216 tick checking mon for new map
debug 2026-02-17T09:50:03.415+0000 7f44b3f23700  1 osd.0 1216 tick checking mon for new map
debug 2026-02-17T09:50:33.539+0000 7f44b3f23700  1 osd.0 1216 tick checking mon for new map
debug 2026-02-17T09:51:04.531+0000 7f44b3f23700  1 osd.0 1216 tick checking mon for new map
debug 2026-02-17T09:51:34.627+0000 7f44b3f23700  1 osd.0 1216 tick checking mon for new map

**Ceph Status from inside Ceph Toolbox pod:**

bash-4.4$ ceph -s
  cluster:
    id:     fa195d4a-3d2a-4c31-8600-1693aa592f7f
    health: HEALTH_WARN
            Reduced data availability: 32 pgs inactive
            11 daemons have recently crashed
            15 slow ops, oldest one blocked for 12475 sec, daemons [mon.c,mon.f,mon.i] have slow ops.

  services:
    mon: 3 daemons, quorum c,f,i (age 3h)
    mgr: a(active, since 16h), standbys: b
    osd: 15 osds: 0 up, 15 in (since 17h)

  data:
    pools:   1 pools, 32 pgs
    objects: 0 objects, 0 B
    usage:   0 B used, 0 B / 0 B avail
    pgs:     100.000% pgs unknown
             32 unknown
             
  Please note that all the Kubernetes OSD pods are running without any crashes.
